### PR TITLE
Remove assert bgzf #3600

### PR DESCRIPTION
**Pull request by vincentdavis**
- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit`` locally,
and understand that AppVeyor and TravisCI will be used to confirm the Biopython unit
tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

issue #3600

### CODE REVIEW

File: Bio/bgzf.py [hunk 1]
```diff
@@ -458,11 +458,14 @@ def _load_bgzf_block(handle, text_mode=False):
         subfield_data = handle.read(subfield_len)
         x_len += subfield_len + 4
         if subfield_id == _bytes_BC:
-            assert subfield_len == 2, "Wrong BC payload length"
-            assert block_size is None, "Two BC subfields?"
+            if subfield_len != 2:
+                raise ValueError("Wrong BC payload length")
+            if block_size is not None:
+                raise AttributeError("Two BC subfields?")
```
Comment (peterjc) [inline Bio/bgzf.py:464]: Why an AttributeError?

**Comment by peterjc** [comment]
Looks good in principle, but why are you using AttributeError in multiple places? It does not seem appropriate to me:
https://docs.python.org/3/library/exceptions.html#AttributeError

**Comment by vincentdavis** [comment]
@peterjc Honestly part of why I wanted to work on this was to learn more about how to choose the correct errors.
I changed a couple to ValueErrors.
I left a couple Attribute errors, I think those are correct.

**Comment by peterjc** [comment]
I don't claim to be an expert on the exceptions, but the two remaining uses of ``AttributeError`` here strike me as better being a ``ValueError`` when validating the method arguments:

```
        if filename and fileobj:
            raise AttributeError("Both filename and fileobj are defined")
```

Happy to call for a 2nd opinion, perhaps @mdehoon or @MarkusPiotrowski ?

**Comment by fabianegli** [comment]
In the case of conflicting arguments I see two possible options:

1. ValueError
2. RuntimeError
    From the python docs:
    > Raised when an error is detected that doesn’t fall in any of the other categories. The associated value is a string indicating what precisely went wrong.

The ValueError is normally raised if ONE value is wrong, but here we are talking about incompatible arguments that can both be correct when supplied to the function without the other. This is why I think ValueError is the wrong choice for the code discussed above and the RuntimeError should be chosen.

**Comment by fabianegli** [comment]
Also `"Both filename and fileobj are defined"` states what's wrong, but not how to use it.

**Comment by peterjc** [comment]
I still think that would be a ``ValueError``, but agree the message should be clearer - e.g. "Supply either filename or fileobj, not both"

**Comment by fabianegli** [comment]
It seems that the stackoverflow consensus is in agreement with @peterjc. Namely to use ValueError for incompatible arguments. Reference: https://stackoverflow.com/a/256260/6018688

**Comment by vincentdavis** [comment]
@peterjc , @fabianegli 
I'll update in the next couple days, thanks for the feedback.

**Comment by peterjc** [comment]
I have resolved the conflict, used the "Supply either filename or fileobj, not both" message, and switched to ValueError.

Assuming all the tests pass, I think we can squash-and-merge this now?

**Comment by fabianegli** [comment]
Random thought: Not sure if there is a biopython standard for the formatting of error messages. The Python standard library normally does not capitalize the first letter of the message and does not use a full stop at the end of the message. I think, however, that if the first letter is capitalized, the message should be ended with a full stop to make it a complete sentence. And vice versa that if there is a full stop at the end of the message it should start capitalized.

**Comment by peterjc** [comment]
I don't think we have ever discussed a consistent standard for formatting our error messages. It seems a nice idea (and could be enforced with a flake8-plugin or something in principle), but technically any change could break people's error handling code so does not seem worth the risk.

**Comment by peterjc** [comment]
Thank you Vincent

**Pull request merged**